key-rotator/key/raw.go

```diff
@@ -0,0 +1,233 @@
+package key
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+)
+
+// Type represents the type of a Raw key.
+type Type uint8
+
+const (
+	// P256 represents an ECDSA P-256 key.
+	P256 Type = 1 + iota
+)
+
+type typeInfo struct {
+	name             string              // string name of type
+	newRandom        func() (raw, error) // function returning a newly-initialized random key
+	newUninitialized func() raw          // function returning an uninitialized key of this type, e.g. for use in unmarshalling
+}
+
+var typeInfos = map[Type]*typeInfo{
+	P256: {"P256", newRandomP256, newUninitializedP256},
+}
+
+func (t Type) String() string {
+	if ti := typeInfos[t]; ti != nil {
+		return ti.name
+	}
+	return "UNKNOWN"
+}
+
+// New creates a new, randomly-initialized key.
+func (t Type) New() (Raw, error) {
+	ti := typeInfos[t]
+	if ti == nil {
+		return Raw{}, fmt.Errorf("unknown key type %v (%d)", t, t)
+	}
+	k, err := ti.newRandom()
+	if err != nil {
+		return Raw{}, fmt.Errorf("couldn't create %v key: %v", t, err)
+	}
+	return Raw{k}, nil
+}
+
+// Raw represents a raw (i.e. unversioned) asymmetric cryptographic key,
+// including both the private & public portions. It has functionality related
+// to serialization of the key.
+type Raw struct{ k raw }
+
+var _ encoding.BinaryMarshaler = Raw{}
+var _ encoding.BinaryUnmarshaler = &Raw{}
+var _ encoding.TextMarshaler = Raw{}
+var _ encoding.TextUnmarshaler = &Raw{}
```

Comment:
Are these just here so that the compiler can enforce that `Raw` implements all four interfaces? If so, could you put in a comment explaining that?

Reply:
Yup, that's why. Comment added.